src/MSBuild/PerformanceLogEventListener.cs

```diff
@@ -0,0 +1,161 @@
+using System;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Text;
+using Microsoft.Build.Eventing;
+
+namespace Microsoft.Build.CommandLine
+{
+    internal sealed class PerformanceLogEventListener : EventListener
+    {
+        internal struct ProviderConfiguration
+        {
+            internal string Name { get; set; }
+            internal EventKeywords Keywords { get; set; }
+            internal EventLevel Level { get; set; }
+        }
+
+        private static ProviderConfiguration[] s_config = new ProviderConfiguration[]
+        {
+            new ProviderConfiguration()
+            {
+                Name = "Microsoft-Build",
+                Keywords = MSBuildEventSource.Keywords.PerformanceLog,
+                Level = EventLevel.Verbose
+            }
+        };
+
+        private const string PerfLogDirEnvVar = "DOTNET_PERFLOG_DIR";
+        private const char EventDelimiter = '\n';
+        private string _processIDStr;
+        private StreamWriter _writer;
+
+        [ThreadStatic]
+        private static StringBuilder s_builder;
+
+        internal static PerformanceLogEventListener Create()
+        {
+            PerformanceLogEventListener eventListener = null;
+            try
+            {
+                // Initialization happens as a separate step and not in the constructor to ensure that
+                // if an exception is thrown during init, we have the opportunity to dispose of the listener,
+                // which will disable any EventSources that have been enabled.  Any EventSources that existed before
+                // this EventListener will be passed to OnEventSourceCreated before our constructor is called, so
+                // we if we do this work in the constructor, and don't get an opportunity to call Dispose, the
+                // EventSources will remain enabled even if there aren't any consuming EventListeners.
+
+                // Check to see if we should enable the event listener.
+                string logDirectory = Environment.GetEnvironmentVariable(PerfLogDirEnvVar);
+                if (Directory.Exists(logDirectory))
+                {
+                    eventListener = new PerformanceLogEventListener();
+                    eventListener.Initialize(logDirectory);
+                }
+            }
+            catch
+            {
+                if (eventListener != null)
+                {
+                    eventListener.Dispose();
```

Comment:
if (eventListener != null && eventListener threw an exception) shouldn't we return null or throw?

Reply:
Yes, that's a good catch.  Thanks.